File: algo/34. Find First and Last Position of Element in Sorted Array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

// @9
// The code in this solution is very robust.
// It works for all scienarios.
// Runtime complexity: Object(log n)

var searchRange = function(nums, target) {
    var ansFirst, ansLast;
    var start = 0;
    var end = nums.length - 1;
    
    while (start + 1 < end) {
        var mid = Math.floor((end - start) / 2) + start;
        if (nums[mid] == target) {
            end = mid;
        } else if (nums[mid] > target) {
            end = mid;
        } else {
            start = mid;
        }
    }
    
    if (nums[end] == target) {
        ansFirst = end;
    }
    
    if (nums[start] == target) {
        ansFirst = start;
    }
    
    if (ansFirst == undefined) {
        ansFirst = -1;
    }
    
    start = 0;
    end = nums.length - 1;
    
    while (start + 1 < end) {
        var mid = Math.floor((end - start) / 2) + start;
        if (nums[mid] == target) {
            start = mid;
        } else if (nums[mid] > target) {
            end = mid;
        } else {
            start = mid;
        }
    }
    
    if (nums[start] == target) {
        ansLast = start;
    }
    
    if (nums[end] == target) {
        ansLast = end;
    }
    
    if (ansLast == undefined) {
        ansLast = -1;
    }
    
    return [ansFirst, ansLast];
    
};

File: round3/34. Find First and Last Position of Element in Sorted Array.cs
public class Solution {
    public int[] SearchRange(int[] nums, int target) {        
        int[] ret = new int[2];
        
        if (nums.Length == 0) {
            ret[0] = -1;
            ret[1] = -1;
            return ret;
        }
        
        int start = 0;
        int end = nums.Length - 1;
        while (start + 1 < end) {
            int mid = (int) (end - start) / 2 + start;
            if (target == nums[mid]) {
                end = mid;
            } else if (target < nums[mid]) {
                end = mid;
            } else {
                start = mid;
            }
        }
        
        if (nums[start] == target) {
            ret[0] = start;
        } else if (nums[end] == target) {
            ret[0] = end;
        } else {
            ret[0] = -1;
            ret[1] = -1;
            return ret;
        }
        
        start = 0;
        end = nums.Length - 1;
        while (start + 1 < end) {
            int mid = (int) (end - start) / 2 + start;
            if (target == nums[mid]) {
                start = mid;
            } else if (target < nums[mid]) {
                end = mid;
            } else {
                start = mid;
            }
        }
        
        if (nums[end] == target) {
            ret[1] = end;
        } else {
            ret[1] = start;
        }
        
        
        return ret;
    }
}